diagnostics-app/src/app/rma/test-viewer/routine-v2-test/routine-v2-test.component.ts

```diff
@@ -0,0 +1,255 @@
+import {Component, EventEmitter, Input, NgZone, Output} from '@angular/core';
+import {LoggingService} from 'app/core/services/logging.service';
+import {RoutineV2Service} from 'app/core/services/routine-v2.service';
+import {
+  TestOrchestratorService,
+  TestResult,
+  V2TestResult,
+} from 'app/core/services/test-orchestrator.service';
+import {
+  RequestType,
+  RoutineV2Argument,
+  RoutineV2Event,
+  RoutineV2EventCategory,
+  RoutineV2FinishedInfoUnion,
+} from 'common/message';
+import {
+  ExceptionInfo,
+  RoutineInitializedInfo,
+  RoutineRunningInfo,
+  RoutineWaitingInfo,
+} from 'common/telemetry-extension-types';
+import {BaseTestComponent} from '../base-test/base-test.component';
+
+enum RoutineV2State {
+  // Either finished or haven't been ran yet.
+  FINISHED = 'finished',
+  INITIALIZED = 'initialized',
+  RUNNING = 'running',
+  WAITING = 'waiting',
+  EXCEPTION = 'exception',
+  CANCELLED = 'cancelled',
+}
+
+@Component({
+  selector: 'app-routine-v2-test',
+  templateUrl: './routine-v2-test.component.html',
+  styleUrls: ['./routine-v2-test.component.css'],
+})
+export class RoutineV2TestComponent implements BaseTestComponent {
+  @Input() index!: number;
+  @Input() title!: string;
+  @Input() argument!: RoutineV2Argument;
+  @Output() testCompleted = new EventEmitter<void>();
+
+  public RoutineV2State = RoutineV2State;
+
+  // The output, which may be either routine result or user action.
+  public output?: object | string;
+  // Whether the routine has passed. Only shown if the routine is finished.
+  public hasPassed?: boolean;
```

Comment:
Why optional? Could we just use `public hasPassed = false`?

Reply:
Changed to hasPassed = false.